instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/internal/JdbcSingletons.java

```diff
@@ -6,43 +6,50 @@
 package io.opentelemetry.instrumentation.jdbc.internal;
 
 import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.OpenTelemetry;
 import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
 import io.opentelemetry.instrumentation.api.instrumenter.SpanKindExtractor;
 import io.opentelemetry.instrumentation.api.instrumenter.db.DbClientSpanNameExtractor;
 import io.opentelemetry.instrumentation.api.instrumenter.db.SqlClientAttributesExtractor;
 import io.opentelemetry.instrumentation.api.instrumenter.net.NetClientAttributesExtractor;
 import io.opentelemetry.instrumentation.api.internal.ConfigPropertiesUtil;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * This class is internal and is hence not for public use. Its APIs are unstable and can change at
  * any time.
  */
 public final class JdbcSingletons {
   public static final String INSTRUMENTATION_NAME = "io.opentelemetry.jdbc";
+  private static final JdbcAttributesGetter dbAttributesGetter = new JdbcAttributesGetter();
+  private static final JdbcNetAttributesGetter netAttributesGetter = new JdbcNetAttributesGetter();
 
-  private static final Instrumenter<DbRequest, Void> INSTRUMENTER;
+  private static final Map<OpenTelemetry, Instrumenter<DbRequest, Void>> instrumenterCache =
+      new ConcurrentHashMap<>();
 
-  static {
-    JdbcAttributesGetter dbAttributesGetter = new JdbcAttributesGetter();
-    JdbcNetAttributesGetter netAttributesGetter = new JdbcNetAttributesGetter();
-
-    INSTRUMENTER =
-        Instrumenter.<DbRequest, Void>builder(
-                GlobalOpenTelemetry.get(),
-                INSTRUMENTATION_NAME,
-                DbClientSpanNameExtractor.create(dbAttributesGetter))
-            .addAttributesExtractor(
-                SqlClientAttributesExtractor.builder(dbAttributesGetter)
-                    .setStatementSanitizationEnabled(
-                        ConfigPropertiesUtil.getBoolean(
-                            "otel.instrumentation.common.db-statement-sanitizer.enabled", true))
-                    .build())
-            .addAttributesExtractor(NetClientAttributesExtractor.create(netAttributesGetter))
-            .buildInstrumenter(SpanKindExtractor.alwaysClient());
+  public static Instrumenter<DbRequest, Void> instrumenter() {
+    return instrumenter(null);
   }
 
-  public static Instrumenter<DbRequest, Void> instrumenter() {
-    return INSTRUMENTER;
+  public static Instrumenter<DbRequest, Void> instrumenter(OpenTelemetry openTelemetry) {
+    if (openTelemetry == null) {
+      openTelemetry = GlobalOpenTelemetry.get();
+    }
+    return instrumenterCache.computeIfAbsent(
```

Comment:
Instead of using this cache, can you turn these `*Singletons` classes into `*InstrumenterFactory` and refactor these methods to simple factory methods?

The `OpenTelemetryDriver` will always pass the `GlobalOpenTelemetry` instance; while the `OpenTelemetryDataSource` should always pass the instance received from the user. There's no need to cache anything, really.